### ISSUE TITLE
2-way binding syntax, e.g., [(ngModel)], should support elvis operator

### ISSUE THREAD

**Issue by wardbell** [angular/angular#7697]
* **I'm submitting a ...**
[ ] bug report
[x] feature request
[ ] support request

* **Do you want to request a *feature* or report a *bug*?**

Feature

* **What is the current behavior?**

A2 supports the elvis operator (?.) in property bindings but not "two-way" bindings.

This is ok: `[ngModel]="hero?.name"`.

A2 throws an error if I write `[(ngModel)]="hero?.name"` because this de-sugars (in part) to `(ngModelChange)="hero?.name=$event"` and the elvis operator (?.) is not supported in assignments.

Today's workaround is  `[ngModel]="hero && hero.name"`. We can do better.

* **If the current behavior is a bug, please provide the steps to reproduce and if possible a minimal demo of the problem** via
https://plnkr.co or similar (you can use this template as a starting point: http://plnkr.co/edit/tpl:AvJOMERrnz94ekVua0u5).

* **What is the expected behavior?**

Recommend that template statements support elvis in assignments.

For example, the following `<input>` element event binding &mdash; `(input)="hero?.name = $event.target.value"` &mdash; should assign the value to `hero.name` if `hero != null`; if `hero == null` it should do nothing.

It follows that we will then support `[(ngModel)]="hero?.name"` which was the original motivation for this request.

* **What is the motivation / use case for changing the behavior?**

Elvis is useful in binding syntax ... both in property binding template expressions and event binding template statements involving assignment.

The 2-way binding syntax today prevents usage of elvis, even if we only wanted it for the property binding half of that syntax.

* **Please tell us about your environment:**

- Angular version: 2.0.0-beta.11 everywhere

**Comment by Foxandxss** [comment]
Oh please, we need this feature. I am a bit annoyed with having to do "foo && foo.bar".

I think it makes sense for Angular 2 template syntax to support some Typescript features. It just need to be documented properly :)

**Comment by PascalPrecht** [comment]
I remember @cburgdorf and me running into this as well. We discussed this with @IgorMinar. 

Igor, if I remember correctly you wanted to discuss this with the team, did you get to that already?

**Comment by nathic** [comment]
still no update on this. 
i would love to see the elvis operator working with 2 way bindings

**Comment by aarmora** [comment]
I also vote for this.  Right now I'm having to initialize every empty model by hand beforehand.

**Comment by ryanlin1986** [comment]
This a essential feature for data-binding

**Comment by wardbell** [comment]
@mhevery You classified as "workaround is obvious". It may be obvious but it's extremely painful to implement across a large app.  Would love to see this get more priority (as would the rest of this gang).

**Comment by PascalPrecht** [comment]
I'd love to hear what came out of the discussions @IgorMinar had about this topics. If I remember correctly, he told us back in the days that this feature is not easily implemented as of now (without performance penalties).

Also, what would be the recommended workaround? We currently preset expressions with some value so we don't need the safe navigation operator. But that's clearly a hack.

**Comment by wardbell** [comment]
I'm not buying the perfect argument for two reasons

1. No one has identified the problem or its performance tax
2. We should have the ability to pay that tax if it makes sense... as it will 99.9% of the time.et is opt in.

To.e for some creativity.

**Comment by cburgdorf** [comment]
If I recall the discussion with @IgorMinar correctly it was more like he wasn't sure if it's the right pattern to promote. 

Promoting `?.` means promoting routing to components that haven't fully loaded yet. It's often better to defer rendering of the component until it's data is ready. 

Personally, I think even if you follow the practice of not rendering components until their data is fully loaded for most of your components you'll still face enough situations where it would be beneficial to allow `?.` with the `[()]` syntax.

**Comment by Jammy42** [comment]
Hi guys - why is the weel reinvented again. In our Company we have moved from Angular 1 to Angular 2. and the current  "null property" behaviour just sucks. (Rewrite all the Code to make sure stuff is pre-initialised).

If the Angular Team is more then just a kindergarden of nerds then I would suggest to provide some backward compatibility (to Angular 1). Otherwise serious players might get p*** off and abandon Angular.

**Comment by Foxandxss** [comment]
@Jammy42 please follow our [Code of Conduct](https://github.com/angular/code-of-conduct/blob/master/CODE_OF_CONDUCT.md) if you plan on contributing to this issue.

Thanks.

**Comment by Jammy42** [comment]
@Foxandxss ... I think you didn't get the "contribution". People out there are spending money and sweat to "follow" the kindergarten. Things like this topics cost a lot (redesign/rewrite) and can even costs jobs - we are not talking about a Hellow-World App, it's about Millions of $ - may be for a young dude like you it's not an issue but there are other ppl out there who think differently.

And finally pls do not argue "it's a free framework" ... it's a framework which has a label Google on it and therefore is also "sold" as such. There are hundreds of template based frameworks out there without and issue about "null" property bindings and as far as I can remember, some folks at Angular  even admit re. this topic to "educate" programmers to become better programmers ... wtf.

**Comment by kenkopelson** [comment]
I really have to take exception to this statement made by cburgdorf above:

"Promoting ?. means promoting routing to components that haven't fully loaded yet. It's often better to defer rendering of the component until it's data is ready."

Why do we use Promises and Observables?  Is it not for this very reason?  WE WANT TO RENDER COMPONENTS NOW and then the data shows up when it shows up.  We are talking javascript here, so when data comes across from someplace like MongoDB, there is NO TYPE to the data.  Fields could easily be missing, since MongoDB does not require empty fields to be in the database.  When fields are missing, I don't expect the binding to break, for heaven's sake.  I expect the binding to say "oh, this field is missing, so I'll just set it anyhow and voila - magically created in the target object!

Please don't presuppose how people will use this and then decide for us what's good or bad.  A framework should be consistent and predictable, making things easier, not harder.  PLEASE FIX THIS NOW, as I consider this a deal-breaker for using Angular 2 at all.  Angular 1 worked very well in this regard.  Please do not take steps backwards.  It's vital we get this fixed.

**Comment by nathic** [comment]
I agree with @kenkopelson and @Jammy42 and I think this shouldn't be marked as feature request. It's much rather a serious bug. In a more and more reactive and async programming world, you can not justify a misfunctional databinding. As @kenkopelson stated: some fields may or may not be set! 
This issue doesn't really bother the angular team and I don't quite understand why. It makes angular 2 really unattractive for to live with the thought of having to preinitialise everything just to make sure not to end up with a completely red console log and a blank site is frustrating and not satisfying. 

For me the workaround is just bull****, do you really think it's a good approach?
For me the label "type: feature" isn't correct for THIS IS A BUG!

I hope you guys can understand my complaints and probably give this a little more attention.

**Comment by Jammy42** [comment]
Would be nice if someone from the Angular2 Team would post a comment here ... or are they already working on a different framework ?

**Comment by vicb** [comment]
@kenkopelson @nathic @Jammy42 please stay constructive and follow our code of conduct.

This is not on our priority list for now as there is a workaround.

Do you know of any langage that supports the safe navigation operator in assignment ? Knowing how other deal with this could help. Thanks.

**Comment by Jammy42** [comment]
Dear @vicb may be you should check the Angular 1 code, where there was **no need** for an Elvis operator ?!

>> priority list for now 
The point here is that obviously your priority list seems not to match with your clients priority list. If the issue wouldn't be an issue then there would be no talking here or am I wrong ?

>> there is a workaround
So you think it's a great service to have your clients producing millions of lines of "workaround" code instead of investing may be a few hours/days to come up with a proper solution ? If yes then you and/or the Angular 2 team do not understand what the idea of a framework is.

I have no problem if there is a timeline for a solution, your client would be able to adjust their plans but right now, one does not know whether to invest days/weeks in producing the "workaround code" or just wait for another RC.

Last but not least - I think Angular is great stuff, but please do not forget your clients, which use it for real (such as eGovernment apps).

**Comment by naomiblack** [comment]
@Jammy42 I'm deleting your comments on this issue and will take further action to remove your ability to comment if necessary.  Bullying and personal attacks on members of the community (calling people out for being young/inexperienced/not "real" team members when you disagree) and demanding action are not tolerated here.   

We have an open issue precisely because we understand that it is valued by many developers.

We have not addressed it because the solution is non-trivial without introducing breaking changes, and there is a workaround that prevents people from being stuck.

I encourage you to review the [code of conduct](https://github.com/angular/code-of-conduct/blob/master/CODE_OF_CONDUCT.md).

**Comment by naomiblack** [comment]
Folks, the discussion on this issue is veering into insults. 

I deleted some comments on this issue that did not add to the technical discussion and I'm locking the conversation for now

Per angular's [code of conduct](https://github.com/angular/code-of-conduct/blob/master/CODE_OF_CONDUCT.md), bullying and personal attacks on members of the community (calling people out for being young/inexperienced/not "real" team members when you disagree) and demanding action are not tolerated here. 

I'd like to assure you that we have an open issue precisely because we understand that it is valued by many developers.

We have not addressed it yet because the solution is non-trivial without introducing breaking changes, and there is a workaround that prevents people from being stuck.

We hear you. It's on our list for future work.

I'm hopeful that we can find a way to add this feature in future, and would welcome contributions that point to other examples of languages that support the safe navigation operator in assignment.

**Comment by petebacondarwin** [comment]
Closing with the same reasoning as given here: https://github.com/angular/angular/issues/42476#issuecomment-902785054
